File: java/src/main/java/com/penapereira/patterns/command/ConcreteCommand.java
package com.penapereira.patterns.command;

/** Binds a receiver to an action and knows how to reverse it. */
public final class ConcreteCommand implements Command {

    private final Receiver receiver;
    private final String word;

    public ConcreteCommand(Receiver receiver, String word) {
        this.receiver = receiver;
        this.word = word;
    }

    @Override
    public void execute() {
        receiver.action(word);
    }

    @Override
    public void undo() {
        receiver.reverse(word);
    }

    @Override
    public String describe() {
        return "ConcreteCommand(" + word + ")";
    }
}

File: typescript/src/command/concrete-command.ts
import type { Command } from "./command.ts";
import type { Receiver } from "./receiver.ts";

/** Binds a receiver to an action and knows how to reverse it. */
export class ConcreteCommand implements Command {
  constructor(
    private readonly receiver: Receiver,
    private readonly word: string,
  ) {}

  execute(): void {
    this.receiver.action(this.word);
  }

  undo(): void {
    this.receiver.reverse(this.word);
  }

  describe(): string {
    return `ConcreteCommand(${this.word})`;
  }
}
